### PR TITLE
add improved std.hash.int - deprecate std.hash.uint32

### PR DESCRIPTION
**Pull request by francescoalemanno**
It turns out that Zig's default bit mixer is _very_ biased, and after a lot of googling it turns out that selecting a better solution is _hard_... 
I wrote a custom statistical analysis taylored for bit mixers in order to select the best one at each size (u64/u32/u16), compared a lot of mixers, and packaged the best ones in this PR.

TESTING CODE [here](https://github.com/francescoalemanno/bit_mixers_analysis)

### CODE REVIEW

**Comment by RetroDev256** [comment]
Thank you for the research! Running both the hash functions in a "counter mode" construction, I tested them using PractRand:

As you can see, the current uint32 fails at 256 MiB:
image
The 32 bit hash function in this PR fails at 1 GiB: (1/4 the state space!)
image
The statistical tests in the PractRand suite are very advanced (even compared to TestU01's BigCrush, which is limited in duration) - I feel this is good evidence we should switch to the new one in this PR.

While less common, I feel that the 64 and 16 bit hash functions would also be nice. One thought though for maintainers, would it be worth it to include a hashUsize function, that switches over @bitSizeOf(usize), selecting one of these functions?

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thank you. Can you please make the function accept the integer type as a parameter, and soft-deprecate the old function (make it keep working but documented as deprecated)?

Just switch on the integer type and put `@compileError("unimplemented")` for the unhandled cases. Or, feel free use the other std.hash abstractions to support those kind of integers.

**Comment by francescoalemanno** [comment]
I believe it is now ready to be merged :)

**Comment by andrewrk** [comment]
Alright this is getting a bit chaotic for me. I'd like to see performance benchmarks on this before it lands. It looks like significantly more instructions now for u32.

**Comment by francescoalemanno** [comment]
Sorry Andrew, I meant well with my sudden modifications :)
I just found out about the ARX (add-rotate-xor) construction, and it seems to be the best tool to devise a general width bit mixer.
I tested it all day using my  method revealing no particular flaws, It also performs much better under PractRand (as suggested by @RetroDev256) .

the excellent news is that it reaches 4 Gigs without failing:

```
➜  bit_mixers_analysis git:(main) ✗ zig run hash_tester.zig -OReleaseFast | RNG_test stdin32    
RNG_test using PractRand version 0.95
RNG = RNG_stdin32, seed = unknown
test set = core, folding = standard (32 bit)

rng=RNG_stdin32, seed=unknown
length= 16 megabytes (2^24 bytes), time= 2.1 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low8/32]BCFN(2+6,13-8,T)         R=  -4.8  p =1-7.8e-5   unusual          
  ...and 119 test result(s) without anomalies

rng=RNG_stdin32, seed=unknown
length= 32 megabytes (2^25 bytes), time= 4.2 seconds
  no anomalies in 131 test result(s)

rng=RNG_stdin32, seed=unknown
length= 64 megabytes (2^26 bytes), time= 8.5 seconds
  no anomalies in 142 test result(s)

rng=RNG_stdin32, seed=unknown
length= 128 megabytes (2^27 bytes), time= 16.7 seconds
  no anomalies in 156 test result(s)

rng=RNG_stdin32, seed=unknown
length= 256 megabytes (2^28 bytes), time= 33.2 seconds
  no anomalies in 168 test result(s)

rng=RNG_stdin32, seed=unknown
length= 512 megabytes (2^29 bytes), time= 65.9 seconds
  no anomalies in 180 test result(s)

rng=RNG_stdin32, seed=unknown
length= 1 gigabyte (2^30 bytes), time= 131 seconds
  no anomalies in 194 test result(s)

rng=RNG_stdin32, seed=unknown
length= 2 gigabytes (2^31 bytes), time= 257 seconds
  no anomalies in 205 test result(s)

rng=RNG_stdin32, seed=unknown
length= 4 gigabytes (2^32 bytes), time= 508 seconds
  no anomalies in 217 test result(s)
  ```
this in turn implies that the properties of this hash are much better, perhaps to an excessive amount... 
If performance is to be preferred over quality, we can revert back to your commit.

**Comment by andrewrk** [comment]
Both are important factors

**Comment by francescoalemanno** [comment]
I do agree with that, the new construction is ~ 12 times slower, but it does not depends on a lot of magic and tuned constants which helps in writing generic code.
Now I am trying to minimise the ARX construction to get a good hash at reasonable speeds.

**Comment by francescoalemanno** [comment]
finally all [requirements](https://ticki.github.io/blog/designing-a-good-non-cryptographic-hash-function/) for a good bit-mixer are met:

- zero sensitivity: 0 is not a fixed point
- diffusion: bias is below 1% at all width except the very
- it is a bijection at all widths, meaning that every int is mapped to a distinct int, it can be used in hash-requiring datastructures without introducing extra collisions.
- performance: it is 1.5 times slower than the original hash in stdlib, but it works at any width and with ~60 times less bias.

@andrewrk, if there are other requirements before merging let me know, and thank you.

**Comment by RetroDev256** [comment]
I would like to add a few thoughts to this PR -

The implementation looks sound - I appreciate the generation of a high quality hash function for all bit widths. I really appreciate the attention to zero-sensitivity and 4 GiB passing for the 32 bit hash function. (really impressive tbh)

I do have two concerns though, albeit minor. The first is that the implementation you suggest seems to be hard to understand. I could understand how bits were manipulated in the original code, but with this new implementation, it is now a bit harder to maintain. Perhaps this could be fixed by performing a hybrid approach? for example - you keep the current int function, and for 16, 32, and 64 bits you call dedicated functions (non-pub) in the current file, which are tailored to provide a balance between speed and quality. Then you can provide a simpler arbitrary-bit length function for the rest of the input integers. With this idea though, I don't feel that hashing 63 or 299 bit integers is really common, so perhaps we may not need this code?

The other concern is a bit minor, you can choose to ignore this - the purpose of zero sensitivity as I know it is to aid programmers that just plop a zero into the function. Most hash functions have a fixed point... or a few thousand. Not sure if this property is worth slowing down the function.

A few thoughts, hope it comes in handy!
I do really think the arbitrary-bit length hash is genius. Better than I could come up with :)

**Comment by andrewrk** [comment]
It looks like you've managed to reveal a bug in the C backend. The next step here will be extracting a reduced behavior test case from this code that triggers the bug and getting that fixed in master branch. Then we can come back and rebase this branch.

**Comment by francescoalemanno** [comment]
Oh okay! I do not have a Linux machine with me currently, is there any way I can help?


Benchmarking strategy
https://github.com/francescoalemanno/bit_mixers_analysis/blob/1cf848fc849489c0055976f436805aaf645d8d30/hash_tester.zig#L81

This benchmarking function times a given hashing function by running it on a range of values (up to 2^32−1), recording the execution time for each run, and keeping track of the best (fastest) observed time. It uses an accumulator, x, to avoid compiler optimizations. After five iterations, it returns the fastest recorded time in nanoseconds.

**Comment by ianprime0509** [comment]
I've reduced the C backend bug and opened it as #21914.

**Comment by francescoalemanno** [comment]
Since the C-backend bug blocked this PR, I figured there was some extra time to squeeze some more performance & quality. I devised some better metrics to evaluate the quality, and now I can see it is pareto optimal.
```
time :      std.hash.uint32               0.1997ns per hash      777566858
time :         std.math.int               0.2800ns per hash      49333260
new hash is 1.4022 times slower.
it has 329 times less bias (evaluated as sqrt(mse_1/mse_2)).
```

The enhanced quality of the filter is also noticed in practice:

immagine

**Comment by RetroDev256** [comment]
I have some bad news - Not completely sure the metrics that are being used, but I feel that the implementation is getting a bit "too smart" for it's original purpose. I won't say that I fully understand the testing, but I feel that we could still do better in terms of quality & speed.

I do think that there may have been some testing accuracy drift over the past few commits, because the current code for 32 bits fails at 1 GiB for me (and 512 MiB is "suspicious" according to PractRand).

I would like to propose a less "autogenerated" hash function. I do understand that it is really nice to be able to automagically hash any integer bit width, but it seems that the current code may be performing sub-par to what I originally reviewed.

I would also like to make a case for lower quality, higher throughput hash functions - If people want a high quality hash function, they can always reach into std.hash for cryptographic hash functions, or a specific construction with high quality output. I feel that the number one use-case for this easy-to-reach hash function will be game & architecture developers who "just want a hash" without all the hastle - which means that speed of the hash may be more important than "perfect hash quality".

This isn't to say that the current ideas behind the implementation are bad or inferior, I just think that the original code I tested may have been better (in terms of use-cases) than the current code.

To satisfy the "make an int function that hashes any bit width", I propose that something along the lines of the following code is introduced:

```zig
pub fn int(x: anytype) @TypeOf(input) {
    switch (@bitSizeOf(input)) {
        0 => return 0,
        1 => return ~x,
        2..8 => return x *% @as(@TypeOf(x), @truncate(0x77)),
        16 => return uint16(x), // non-pub decl in current file
        32 => return uint32(x), // eventually non-pub decl in current file
        64 => return uint64(x), // non-pub decl in current file
        else => @compileError("int() function not implemented for current type"),
    };
}
```

<details>
<summary>Current PR Hash Function Test Results</summary>

```
ramen:[retrodev]:~/repos/Zig/test$ zig build run -Doptimize=ReleaseFast | RNG_test stdin32
RNG_test using PractRand version 0.95
RNG = RNG_stdin32, seed = unknown
test set = core, folding = standard (32 bit)

rng=RNG_stdin32, seed=unknown
length= 256 megabytes (2^28 bytes), time= 3.3 seconds
  no anomalies in 168 test result(s)

rng=RNG_stdin32, seed=unknown
length= 512 megabytes (2^29 bytes), time= 8.3 seconds
  Test Name                         Raw       Processed     Evaluation
  FPF-14+6/16:all                   R=  -5.8  p =1-3.2e-5   mildly suspicious
  ...and 179 test result(s) without anomalies

rng=RNG_stdin32, seed=unknown
length= 1 gigabyte (2^30 bytes), time= 15.2 seconds
  Test Name                         Raw       Processed     Evaluation
  FPF-14+6/16:(0,14-0)              R=  -9.1  p =1-4.1e-8   suspicious       
  FPF-14+6/16:(1,14-0)              R=  -7.3  p =1-2.2e-6   unusual          
  FPF-14+6/16:all                   R= -13.8  p =1-3.8e-13    FAIL           
  ...and 191 test result(s) without anomalies
  ```

</details>

<details>
<summary>Current PR Hash Function Testing Code</summary>

```zig
const std = @import("std");

pub fn main() !void {
    const endian = @import("builtin").cpu.arch.endian();

    const stdout = std.io.getStdOut().writer();
    var bw = std.io.bufferedWriter(stdout);
    const writer = bw.writer();

    var input: u32 = 0;
    while (true) : (input +%= 1) {
        const hash = int(input);
        writer.writeInt(u32, hash, endian) catch unreachable;
    }

    try bw.flush();
}

/// Applies a bit-mangling transformation to an integer type `T`.
pub fn int(input: anytype) @TypeOf(input) {
    const int_type = @TypeOf(input);
    const bits = @typeInfo(int_type).int.bits;
    const Unsigned = @Type(.{ .int = .{ .signedness = .unsigned, .bits = bits } });
    var x: Unsigned = @bitCast(input);
    if (bits == 0) {
        x = 0;
    } else if (bits == 1) {
        x = ~x;
    } else if (bits == 2) {
        const d = x +% 1;
        x = ~((d >> 1) | (d << 1));
    } else {
        const constants = comptime blk: {
            const shift_1 = (bits >> 1);
            const shift_2 = shift_1 + 1;
            var rng = @import("std").Random.Pcg.init(0xc8dea3c3eab7eaa1);
            const sampler = rng.random();
            const v = .{
                .add = sampler.int(Unsigned),
                .mulshifts = .{
                    .{ sampler.int(Unsigned) | 1, shift_2 },
                    .{ sampler.int(Unsigned) | 1, shift_1 },
                    .{ sampler.int(Unsigned) | 1, shift_2 },
                    .{ sampler.int(Unsigned) | 1, shift_1 },
                },
            };
            break :blk v;
        };
        x +%= constants.add;
        inline for (constants.mulshifts) |ms| {
            x *%= ms[0];
            x ^= x >> ms[1];
        }
    }
    return @bitCast(x);
}
```

</details>

**Comment by andrewrk** [comment]
Reminder that I marked an earlier state of this branch as auto-merge. I think it was a mistake to force-push again rather than wait until that landed and propose a new change.

**Comment by francescoalemanno** [comment]
@RetroDev256 : past PR and current PR failt practrand at 1GiB, still 4 times bigger than stdlib current default. the metric used is to minimise 1 point correlation function and 2 points correlation functions over the variable "hash(r)^hash(r xor (1<<b))", r,b random in their respective ranges. these two metrics suffice to ensure uniformity of diffusion and unbiasedness (the qualities needed for a non-crypto hash). 

@andrewrk : Sorry, I was under the impression that after merging it would become unchangeable for a very long time. And I felt that it could be improved.

In any case, I am in favor of reverting to a more controllable implementation, can I just put a few tweaks on top of RetroDev proposal (which is the natural evolution of the branch that was to be automerged... sorry guys) and move forward?

**Comment by RetroDev256** [comment]
Thank you for this PR @francescoalemanno, I really like the effort you put into this. LGTM.

File: lib/std/hash.zig
Comment (andrewrk) [inline lib/std/hash.zig:44]:
```suggestion
    const info = @typeInfo(@TypeOf(input)).int;
```

I did this already in 5d790e911fbca8bd04392bddad500b96a1837d9c

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: lib/std/hash.zig
Comment (francescoalemanno) [reply lib/std/hash.zig]: okay done.
is this preferred because a good error message if type != int at some point will be automatically printed?

File: lib/std/hash.zig
Comment (mlugg) [inline lib/std/hash.zig:82]: This should be a doc comment. The other `Source:` comments you've added also probably should be, but this one is particularly important since it's a deprecation notice.

File: lib/std/hash.zig
Comment (mlugg) [reply lib/std/hash.zig]: Perhaps the compiler will get "smart" enough to emit a more specific error message here, or perhaps not; but the standard library shouldn't be littered with manual type checks / explicit compile errors when the error you'd otherwise get is relatively understandable, as this one is. Generally speaking, `@compileError` shouldn't be used too often.

Comment (francescoalemanno) [reply lib/std/hash.zig]: ok thank you!

**Comment by francescoalemanno** [comment]
can this be merged?

**Comment by NoReligion** [comment]
The heuristic construction could be expanded to improve consistency.

```zig
else => output: {
    const shift = bits / 2;
    var x: Uint = u_input;
    x = (x ^ (x >> shift)) *% mult;
    x = (x ^ (x >> shift)) *% mult;
    x = (x ^ (x >> shift)) *% mult;
    break :output x ^ (x >> shift);
},
```

**Comment by francescoalemanno** [comment]
> The heuristic construction could be expanded to improve consistency.
> 
> ```zig
> else => output: {
>     const shift = bits / 2;
>     var x: Uint = u_input;
>     x = (x ^ (x >> shift)) *% mult;
>     x = (x ^ (x >> shift)) *% mult;
>     x = (x ^ (x >> shift)) *% mult;
>     break :output x ^ (x >> shift);
> },
> ```

it is not equal, you are missing a mul, also the code has already been reviewed by both andrewrk and mlugg.

**Comment by francescoalemanno** [comment]
Guys can this PR be merged?
@andrewrk @mlugg

File: lib/std/hash.zig
Comment (andrewrk) [inline lib/std/hash.zig:40]: Please avoid words like "easy" and "fast" in doc comments. These are subjective feelings that do not belong in technical writing.

```suggestion
/// Integer to integer hashing for bit widths <= 256.
```

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thanks for the follow-ups. I've made one requested change that I will commit locally and force-push to your branch. Please leave it to me and I will get this landed.

**Comment by andrewrk** [comment]
I'm omitting this from the release notes because I don't see any clean copy pastable content here

**Pull request merged**
